zephyr/lib/alloc.c

```diff
@@ -189,6 +196,86 @@ static void l3_heap_free(struct k_heap *h, void *mem)
 
 #endif
 
+#if CONFIG_VIRTUAL_HEAP
+static void *virtual_heap_alloc(struct vmh_heap *heap, uint32_t flags, uint32_t caps, size_t bytes,
+				uint32_t align)
+{
+	void *mem;
+
+	K_SPINLOCK(&vmh_lock) {
+		heap->core_id = cpu_get_id();
+		mem = vmh_alloc(heap, bytes);
+	}
+
+	if (!mem)
+		return NULL;
+
+	assert(IS_ALIGNED(mem, align));
+
+	if (flags & SOF_MEM_FLAG_COHERENT)
+		return sys_cache_uncached_ptr_get((__sparse_force void __sparse_cache *)mem);
+
+	return mem;
+}
+
+/**
+ * Checks whether pointer is from virtual memory range.
+ * @param ptr Pointer to memory being checked.
+ * @return True if pointer falls into virtual memory region, false otherwise.
+ */
+static bool is_virtual_heap_pointer(void *ptr)
+{
+	uintptr_t virtual_heap_start = POINTER_TO_UINT(sys_cache_cached_ptr_get(&heapmem)) +
+				       HEAPMEM_SIZE;
+	uintptr_t virtual_heap_end = CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE;
+
+	if (!is_cached(ptr))
+		ptr = sys_cache_cached_ptr_get((__sparse_force void *)ptr);
+
+	return ((POINTER_TO_UINT(ptr) >= virtual_heap_start) &&
+		(POINTER_TO_UINT(ptr) < virtual_heap_end));
+}
+
+static void virtual_heap_free(void *ptr)
+{
+	ptr = sys_cache_cached_ptr_get((__sparse_force void *)ptr);
+
+	K_SPINLOCK(&vmh_lock) {
+		vmh_free(virtual_buffers_heap, ptr);
+	}
+}
+
+static const struct vmh_heap_config static_hp_buffers = {
```

Comment:
Btw, since we have a MAX size macros for this, we should probably use it here in the declaration. It will probably help the CC and static analysis tool do bounds checking.

Reply:
@lgirdwood Exactly what macro are you referring to and where would it be used?

Reply:
@softwarecki sorry, I meant ```MAX_MEMORY_ALLOCATORS_COUNT```.  I'm assuming this is the size and not something else ?

Reply:
@lgirdwood The `struct vmh_heap_config` already contains `MAX_MEMORY_ALLOCATORS_COUNT` array of `struct vmh_block_bundle_descriptor`.

```C
struct vmh_heap_config {
	struct vmh_block_bundle_descriptor block_bundles_table[MAX_MEMORY_ALLOCATORS_COUNT];
};
```